File: benchmarks/python/heap_find_min.py
import random
import pprint

class node:
	def __init__(self, key = None, order = None, children = None, sibling = None):
		self.key = key
		self.order = order
		self.children = children
		self.sibling = sibling

#############
# acc func  #
#############
def create_heap(size):
	root = node()
	for i in range(1, size):
		new_node = node(random.randint(0, 20), 0, None, None)
		root = merge(root, new_node)

def merge(x, y):
	if x is None:
		return y
	elif y is None:
		return x
	elif x.order == y.order:
		xs = x.sibling
		ys = y.sibling
		z = merge(xs, ys)
		ord1 = x.order
		ord2 = y.order
		if ord1 <= ord2:
			xchld = x.children
			y.sibling = xchld
			x.children = ys
			x.sibling = z
			x.order = ord1 + 1
			return x
		else:
			ychld = y.children
            x.sibling = ychld
            y.children = x
            y.sibling = z
            y.order = ord2 + 1
            return y
	elif y.order < x.order:
		xs = x.sibling
		z = merge(xs, y)
		ord1 = x.order
		ord2 = y.order
		zord = z.order
		zk = z.key
		xk = x.key

		if ord1 == zord and zk <= xk:
			zchld = z.children
			x.sibling = zchld
			z.children = x
			z.order = zord + 1
			return z
		elif ord1 == zord and xk < zk:
			zsib = z.sibling
			x.sibling = zsib
			xchld = x.children
			z.sibling = xchld
			x.children = z
			x.order = ord1 + 1
			return x
		else:
			x.sibling = z
			return x
	else:
		ys = y.sibling
		z = merge(ys, x)
		ord1 = x.order
		ord2 = y.order
		zord = z.order
		zk = z.key
		yk = y.key
		if ord2 == zord and zk <= yk:
			zchld = z.children
			y.sibling = zchld
			z.children = y
			z.order = zord + 1
			return z
		elif ord2 == zord and yk < zk:
			zsib = z.sibling
			y.sibling = zsib
			ychld = y.children
			z.sibling = ychld
			y.children = z
			y.order = ord2 + 1
			return x
		else:
			y.sibling = z
			return y
#############
# end acc   #
#############

def find_min(x):
	xs = x.sibling
	xk = x.key
	if xs is None:
		return xk
	else
	y = find_min(xs)
	if xk <= y:
		return xk
	else:
		return y

def main():
	x = create_heap(5)
	find_min(x)

if __name__=="__main__":
	main()

File: benchmarks/java/retrieve heap/src/testsuites/Heap_Find_Min.java
package testsuites;

import java.util.Random;

public class Heap_Find_Min {
    static SNnode create_heap(int size){
        Random rand = new Random();
        int i = 0;
        SNnode root = null;
        for(i = 0; i < size; i++){
            SNnode x = new SNnode();
            x.key = rand.nextInt(20);
            x.order = 0;
            x.children = null;
            x.sibling = null;
            root = binomial_heap_merge_rec(root, x);
        }
        return root;
    }

    static SNnode binomial_heap_merge_rec(SNnode x, SNnode y)
    {
        if (x == null) {
            return y;
        }
        else if (y == null) {
            return x;
        }
        else if (x.order == y.order) {
            SNnode xs = x.sibling;
            SNnode ys = y.sibling;
            SNnode z = binomial_heap_merge_rec(xs, ys);
            int ord1 = x.order;
            int ord2 = y.order;
            if (ord1 <= ord2) {
                SNnode xchld = x.children;
                y.sibling = xchld;
                x.children = y;
                x.sibling = z;
                x.order = ord1 + 1;
                return x;
            } else {
                SNnode ychld = y.children;
                x.sibling = ychld;
                y.children = x;
                y.sibling = z;
                y.order = ord2 + 1;
                return y;
            }
        }
        else if (y.order < x.order) {
            SNnode xs = x.sibling;
            SNnode z = binomial_heap_merge_rec(xs, y);
            int ord1 = x.order;
            int ord2 = y.order;
            int zord = z.order;
            int zk = z.key;
            int xk = x.key;

            if (ord1 == zord && zk <= xk) {
                SNnode zchld = z.children;
                x.sibling = zchld;
                z.children = x;
                z.order = zord + 1;
                return z;
            } else if (ord1 == zord && xk < zk) {
                SNnode zsib = z.sibling;
                x.sibling = zsib;
                SNnode xchld = x.children;
                z.sibling = xchld;
                x.children = z;
                x.order = ord1 + 1;
                return x;
            } else {
                x.sibling = z;
                return x;
            }
        }
        else {
            SNnode ys = y.sibling;
            SNnode z = binomial_heap_merge_rec(ys, x);
            int ord1 = x.order;
            int ord2 = y.order;
            int zord = z.order;
            int zk = z.key;
            int yk = y.key;

            if (ord2 == zord && zk <= yk) {
                SNnode zchld = z.children;
                y.sibling = zchld;
                z.children = y;
                z.order = zord + 1;
                return z;
            } else if (ord2 == zord && yk < zk) {
                SNnode zsib = z.sibling;
                y.sibling = zsib;
                SNnode ychld = y.children;
                z.sibling = ychld;
                y.children = z;
                y.order = ord2 + 1;
                return x;
            } else {
                y.sibling = z;
                return y;
            }
        }
    }

    static int binomial_heap_find__minimum_rec(SNnode x)
    {
        //pre
        SNnode xs = x.sibling;
        int xk = x.key;

        if (xs == null) {
            //post
            return xk;
        } else {
            int y = binomial_heap_find__minimum_rec(xs);
            if (xk <= y)
                //post
                return xk;
            else
                //post
                return y;
        }
    }
    
    public static void main(String args[]){
        SNnode x = null;
        x = create_heap(10);
        int res = binomial_heap_find__minimum_rec(x);
    }
}

class SNnode{
    int key;
    int order;
    SNnode children;
    SNnode sibling;
}
